corehq/ex-submodules/casexml/apps/phone/tests/test_cleanliness.py

```diff
@@ -0,0 +1,108 @@
+                    CaseRelationship(),
+                ]
+            )
+        )
+        self.assert_owner_clean()  # this is an actual assertion
+
+    @property
+    def owner_cleanliness(self):
+        return OwnershipCleanliness.objects.get_or_create(
+            owner_id=self.owner_id,
+            domain=self.domain,
+            defaults={'is_clean': True}
+        )[0]
+
+    def assert_owner_clean(self):
+        return self.assertTrue(self.owner_cleanliness.is_clean)
+
+    def assert_owner_dirty(self):
+        return self.assertFalse(self.owner_cleanliness.is_clean)
+
+    def _set_owner(self, case_id, owner_id):
+        case = self.factory.create_or_update_case(
+            CaseStructure(case_id=case_id, attrs={'create': False, 'owner_id': owner_id})
+        )[0]
+        self.assertEqual(owner_id, case.owner_id)
+
+    def test_add_normal_case_stays_clean(self):
+        # change the owner ID of a normal case, should remain clean
+        self.factory.create_case()
+        self.assert_owner_clean()
+
+    def test_change_owner_stays_clean(self):
+        # change the owner ID of a normal case, should remain clean
+        new_owner = uuid.uuid4().hex
+        self._set_owner(self.sample_case._id, new_owner)
+        self.assert_owner_clean()
+
+    def test_change_owner_child_case_stays_clean(self):
+        # change the owner ID of a child case, should remain clean
+        new_owner = uuid.uuid4().hex
+        self._set_owner(self.child._id, new_owner)
+        self.assert_owner_clean()
+
+    def test_add_clean_parent_stays_clean(self):
+        # add a parent with the same owner, should remain clean
+        self.factory.create_or_update_case(CaseStructure(relationships=[CaseRelationship()]))
+        self.assert_owner_clean()
+
+    def test_create_dirty_makes_dirty(self):
+        # create a case and a parent case with a different owner at the same time
+        # make sure the owner becomes dirty.
+        new_owner = uuid.uuid4().hex
+        [child, parent] = self.factory.create_or_update_case(
+            CaseStructure(
+                relationships=[
+                    CaseRelationship(CaseStructure(attrs={'owner_id': new_owner}))
+                ]
+            )
+        )
+        self.assert_owner_dirty()
+        self.assertEqual(parent._id, self.owner_cleanliness.hint)
+
+    def test_add_dirty_parent_makes_dirty(self):
+        # add parent with a different owner and make sure the owner becomes dirty
+        new_owner = uuid.uuid4().hex
+        [child, parent] = self.factory.create_or_update_case(
+            CaseStructure(
+                case_id=self.sample_case._id,
+                relationships=[
+                    CaseRelationship(CaseStructure(attrs={'owner_id': new_owner}))
+                ]
+            )
+        )
+        self.assert_owner_dirty()
+        self.assertEqual(parent._id, self.owner_cleanliness.hint)
+
+    def test_change_parent_owner_makes_dirty(self):
```

Comment:
:+1: yes that makes sense

On 2 April 2015 at 14:08, Cory Zue <notifications@github.com> wrote:

> In corehq/ex-submodules/casexml/apps/phone/tests/test_cleanliness.py
> <https://github.com/dimagi/commcare-hq/pull/6149#discussion_r27651033>:
>
> > +
> > +    def test_add_dirty_parent_makes_dirty(self):
> > +        # add parent with a different owner and make sure the owner becomes dirty
> > +        new_owner = uuid.uuid4().hex
> > +        [child, parent] = self.factory.create_or_update_case(
> > +            CaseStructure(
> > +                case_id=self.sample_case._id,
> > +                relationships=[
> > +                    CaseRelationship(CaseStructure(attrs={'owner_id': new_owner}))
> > +                ]
> > +            )
> > +        )
> > +        self.assert_owner_dirty()
> > +        self.assertEqual(parent._id, self.owner_cleanliness.hint)
> > +
> > +    def test_change_parent_owner_makes_dirty(self):
>
> Good call - i remember that conversation now. I think that your race
> condition isn't a problem, because the question we can ask is "did this
> case trigger a cleanliness conflict? if yes, set the dirtiness, otherwise
> do nothing". then it's totally consistent for each form.
>
> I'll play with this today or tomorrow and see if i can get it working
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/6149/files#r27651033>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa